README
libconfini {#readme}
====================

**libconfini** is the ultimate and most consistent INI file parser library
written in C. Originally designed for parsing configuration files written by
other programs, it focuses on standardization and parsing exactness and is at
ease with almost every type of file containing key/value pairs.

The library is fast and suitable for embedded systems. Its algorithms are
written from scratch and do not depend on any external library, except for the
C standard headers **stdio.h**, **stdlib.h**, **stdbool.h** and **stdint.h**
(and for extreme platforms the code [can be also compiled as “bare metal”][1],
with few or no strings attached to the C Standard Library).

Rather than storing the parsed data, **libconfini** gives the developer the
freedom to choose what to do with them through a custom callback invoked for
each INI node read. The API has been designed to be powerful, flexible and
simple to use.

With **libconfini** you will find in INI files the same serialization power
that you would normally find in other heavily structured formats (such as JSON,
YAML, TOML), but with the advantage of using the most human-friendly
configuration format ever invented (thanks to their informal status, INI files
are indeed more fluid, expressive and human-friendly than formats explicitly
designed with the same purpose, such as YAML and TOML). The library's main goal
is to be uncommonly powerful in the most tedious and error-prone task when
parsing a text file in C: string handling. Thanks to this, the programmer is
left free to organize the content parsed as (s)he pleases, assisted by a rich
set of fine-tuning tools.


Features
--------

* Typed data support (each value can be parsed as a boolean, a number, a
  string, an array)
* Single/double quotes support in _Bash style_ (i.e. quotes can be opened and
  closed multiple times within the same value)
* Multi-line support
* Comment support
* Disabled entry support
* INI sectioning support (single-level sectioning, as in `[foo]`; absolute
  nesting, as in `[foo.bar]`; relative nesting, as in `[.bar]`)
* Automatic sanitization of values, key names and section paths
* Comparison functions designed just for INI source code (capable, for example,
  to recognize the equality between <code>"Hello world"</code> and
  <code>"He"l'lo' world</code>, or between <code>foo bar</code> and
  <code>foo&nbsp;&nbsp;&nbsp;&nbsp;bar</code>)
* Callback pattern
* Thread-safety (each parsing process is fully reentrant)
* Highly optimized code (single memory allocation for each parsing, heuristic
  programming, optimization flags)
* Function modularity (each public function is independent from all other
  public functions)
* K.I.S.S. (no public functions are automatically invoked during the parsing --
  for example, not even single/double quotes are automatically removed from
  values unless the developer explicitly decides to use the formatting
  functions provided by the API)
* Robust and cross-platform file access (UTF-8 support; protection against null
  byte injection; support of all code representations of new lines -- i.e.
  ubiquitous support of Classic Mac OS' `CR`, Unix' `LF`, Windows' `CR`+`LF`,
  RISC OS Open's `LF`+`CR`)


Sample usage
------------

log.ini:

`````````````````````````````````````` ini
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
``````````````````````````````````````

example.c:

``````````````````````````````````````````````````````````````````````````` c
#include <confini.h>

static int callback (IniDispatch * disp, void * v_other) {

  #define IS_KEY(SECTION, KEY) \
    (ini_array_match(SECTION, disp->append_to, '.', disp->format) && \
    ini_string_match_ii(KEY, disp->data, disp->format))

  if (disp->type == INI_KEY) {

    if (IS_KEY("users", "have_visited")) {

      /* No need to parse this field as an array right now */
      printf("People who have visited: %s\n", disp->value);

    } else if (IS_KEY("last_update", "date")) {

      printf("Last update: %s\n", disp->value);

    }

  }

  #undef IS_KEY

  return 0;

}

int main () {

  if (load_ini_path("log.ini", INI_DEFAULT_FORMAT, NULL, callback, NULL)) {

    fprintf(stderr, "Sorry, something went wrong :-(\n");
    return 1;

  }

  return 0;

}
```````````````````````````````````````````````````````````````````````````

Output:

````````````````````````````````````````````````
People who have visited: ronnie, lilly82, robrob
Last update: 12.03.2017
````````````````````````````````````````````````

If you are using C++, under `examples/cplusplus/map.cpp` you can find a snippet
for storing an INI file into a class that relies on a `std::unordered_map`
object.

For more details, please read the [Library Functions Manual][2] (`man
libconfini`) -- a standalone HTML version is available [here][3] -- and the
manual of the header file (`man confini.h`). The code is available on
[GitHub][4] under [madmurphy/libconfini][5]).


Installation
------------

Despite the small footprint, **libconfini** has been conceived as a shared
library (but it can be used as a static library as well). An automatic list of
the distributions that ship the library already compiled is available [here][6].

If a pre-compiled package for your platform is not available, on most Unix-like
systems it is possible to install **libconfini** using the following common
steps:

`````````````````` sh
./configure
make
make install-strip
``````````````````

If the `strip` utility is not available on your machine, use `make install`
instead (it will produce larger binaries)

For a minimum installation without development files (i.e. static libraries,
headers, documentation, examples, etc.) use `./configure --disable-devel`.

If the `configure` script is missing you need to generate it by running the
`bootstrap` script. By default, `bootstrap` will also run the `configure`
script immediately after having generated it, so you may type the `make`
command directly after `bootstrap`. To list different options use `./bootstrap
--help`.

If you are using **Microsoft Windows**, a batch script for compiling
**libconfini** with [**MinGW**][7] without **Bash** is available
(`mgwmake.bat`). If you are interested in using **GNU Make** for compiling
**libconfini**, you can integrate **MinGW** with [**MSYS**][8], or you can
directly use [**MSYS2**][9] and [its official port][10] of the library.
Alternatively, [an unofficial port][11] of **libconfini** for [**Cygwin**][12]
is available.

For further information, see [INSTALL][13].


Bindings
--------

[Danilo Spinella][14] maintains [a D binding package][15] of **libconfini**.


Free software
-------------

This library is free software. You can redistribute it and/or modify it under
the terms of the GPL license version 3 or any later version. See [COPYING][16]
for details.


  [1]: https://madmurphy.github.io/libconfini/html/baremetal.html
  [2]: https://madmurphy.github.io/libconfini/html/libconfini.html
  [3]: https://madmurphy.github.io/libconfini/manual.html
  [4]: https://github.com/
  [5]: https://github.com/madmurphy/libconfini
  [6]: https://repology.org/project/libconfini/versions
  [7]: http://www.mingw.org/
  [8]: http://www.mingw.org/wiki/MSYS
  [9]: https://www.msys2.org/
  [10]: https://packages.msys2.org/base/mingw-w64-libconfini
  [11]: https://github.com/fd00/yacp/tree/master/libconfini
  [12]: https://www.cygwin.com/
  [13]: https://madmurphy.github.io/libconfini/html/install.html
  [14]: https://danyspin97.org/
  [15]: https://github.com/DanySpin97/libconfini-d
  [16]: https://github.com/madmurphy/libconfini/blob/master/COPYING



NEWS
News
====


## 1.16.4

* A bug has been fixed in `load_ini_path()` (`fclose()` was not invoked in case of
  errors)


## 1.16.2

* Issue [#16](https://github.com/madmurphy/libconfini/issues/16) has been fixed


## 1.16.1

* A batch script for compiling **libconfini** under **Microsoft Windows** with
  **MSVC** has been created (`msvcmake.bat`)


## 1.16.0

* Function `ini_get_bool_i()` has been created
* New examples have been created (`examples/cplusplus/map.cpp`,
  `examples/utilities/ini_string_preparse.h`)


## 1.15.0

* Public function `ini_global_set_lowercase_mode()` and global variable
  `INI_GLOBAL_LOWERCASE_MODE` have been marked as deprecated
* Global integer constant `INI_DISABLED_FLAG` and macro
  `INI_IS_IMPLICIT_SUBSTR()` have been added to the public header


## 1.14.0

* Added implicit value check in `ini_array_break()`, `ini_array_collapse()`,
  `ini_array_release()`, `ini_array_split()`, `ini_string_parse()` and
  `ini_unquote()`
* Integer constant `CONFINI_EROADDR` has been added to the `ConfiniInterruptNo`
  `enum`
* New examples have been created (`examples/miscellanea/parse_foreign.c`,
  `examples/miscellanea/toml-like.c` and
  `examples/utilities/clone_ini_dispatch.h`)


## 1.13.0

* Standard header `stdbool.h` has been added to the library's public header for
  handling C++ compatibility
* Module-definition file `src/libconfini.def` has been added to the package
  tree for improving compatibility with **Microsoft Windows**
* A `--without-io-api` option has been added to the `configure` script for
  compiling the library without the C Standard I/O API
* **GNU Autotests** have been implemented (via `make check`, `make finishcheck`
  and `make installcheck` -- launch `make help` for more information about
  these targets)


## 1.12.0

* Function `ini_get_float()` has been marked as deprecated for parsing a
  `double` data type: use `ini_get_double()` instead


## 1.11.0

* Literal constants `CONFINI_EBADF` and `CONFINI_EFBIG` have been added to
  the `ConfiniInterruptNo` `enum`
* POSIX I/O API is now preferably used when available
* A `--with-io-api=API` option has been added to the `configure` script for
  selecting the I/O API to use) 


## 1.10.4

* Added a `--disable-devel` option to the `configure` script for generating
  binary-only installations -- i.e. without installing headers, static
  libraries, documentation, examples, etc.)
* Added call to `strip.exe` to the `mgwmake.bat` script


## 1.10.2

* New examples have been created (`examples/utilities/load_ini_buffer.h` and
  `examples/utilities/make_strarray.h`)
* The `configure` script has been improved with a built-in support for complete
  package renaming (using `--program-prefix=PREFIX`, `--program-suffix=SUFFX`
  and `--program-transform-name='s/[OLDTEXT]/[NEWTEXT]/[g]'`) and several new
  options (`--enable-version-info` and `--disable-version-info`, for
  enabling/disabling **libtool** versioning system; `--enable-extended-config`,
  for safely exporting package renamings at `configure` time, or any changes in
  `configure.ac`, directly to `src/winres.rc` and `package.json` -- see
  `INSTALL` and `./configure --help` for more information)


## 1.10.0

* Function `strip_ini_cache()` has been created
* New examples have been created (`examples/topics/strip_ini_cache.c`)


## 1.9.2

* New examples have been created (`examples/miscellanea/colon_as_delimiter.c`)


## 1.9.0

* Added † `--multiversion` option to the † `autogen.sh` script for future major
  version changes (see `INSTALL`)


## 1.8.0

* Introduced **libtool** versioning system for the compiled binary (versioning
  begins with current version 1.8.0, represented as `0:0:0`); note that
  **libtool** versioning system does not affect the versioning of the project
  currently in use, but constitutes only a means for the linker to keep track
  of incompatibilities between different releases of the library
* New examples have been created (`examples/topics/ini_array_foreach.c`,
  `examples/topics/ini_array_split.c` and `examples/miscellanea/stats_only.c`)


## 1.7.0

* Default format `INI_UNIXLIKE_FORMAT` has been created


## 1.6.1 (unstable)

* Function `ini_array_match()` has been created


## 1.6.0 (unstable)

* Functions `ini_array_shift()`, `ini_array_break()`, `ini_array_release()`
  have been created
* Boolean pair `OFF`/`ON` has been added to the list of supported INI booleans
* Boolean pair `0`/`1` has been removed from the list of supported INI booleans
* Redundant function † `ini_get_lazy_bool()` has been removed


## 1.5.2 (unstable)

* Option `IniFormat::section_paths` has been added to the `IniFormat` bitfield


## 1.5.0 (unstable)

* Added `IniFormat::preserve_empty_quotes` option to the `IniFormat` bitfield
* Function `ini_string_parse()` has been created
* Default value of `INI_GLOBAL_LOWERCASE_MODE` has been set to `FALSE`
* Made sure that sections appended to root as pseudo-subsections will be
  dispatched always without the leading dot
* Made sure that key and section names will always be cleaned of empty quotes
  in formats that support quotes (relevant for space collapsing)

Due to the many renamings this release is largely incompatible with previous
versions. However it is possible to reach full backward compatibility with
version 1.4.2 by using the following aliases:

    #define hash hash_marker
    #define semicolon semicolon_marker
    #define multiline_entries multiline_nodes
    #define ini_collapse_array ini_array_collapse
    #define ini_split_array ini_array_split
    #define ini_set_implicit_value ini_global_set_implicit_value
    #define INI_INSENSITIVE_LOWERCASE INI_GLOBAL_LOWERCASE_MODE
    #define IniFormatId IniFormatNum
    #define ini_format_get_id ini_fton
    #define ini_format_set_to_id(DEST, SRC) (*(DEST) = ini_ntof(SRC))
    #define INI_PARSE_COMMENT INI_DISABLED_OR_COMMENT
    #define INI_SHOW_COMMENT INI_ONLY_COMMENT
    #define INI_FORGET_COMMENT INI_IGNORE

src/confini.h
/*  -*- Mode: C; indent-tabs-mode: nil; c-basic-offset: 4; tab-width: 4 -*-  */

/**

    @file       confini.h
    @brief      libconfini header
    @author     Stefano Gioffr&eacute;
    @copyright  GNU General Public License, version 3 or any later version
    @version    1.16.4
    @date       2016-2022
    @see        https://madmurphy.github.io/libconfini

**/


#ifndef __CONFINI_H__
#define __CONFINI_H__



#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>



#ifdef __cplusplus
extern "C" {
#endif



/*  PRIVATE (HEADER-SCOPED) MACROS  */


#ifdef _LIBCONFINI_NOCCWARN_
#define _LIBCONFINI_DO_PRAGMA_(PBODY)
#define _LIBCONFINI_WARNING_(MSG)
#elif defined(_MSC_VER)
#define _LIBCONFINI_DO_PRAGMA_(PBODY) __pragma(PBODY)
#define _LIBCONFINI_WARNING_(MSG) _LIBCONFINI_DO_PRAGMA_(message("*WARNING*: " MSG))
#else
#define _LIBCONFINI_DO_PRAGMA_(PBODY) _Pragma(#PBODY)
#define _LIBCONFINI_WARNING_(MSG) _LIBCONFINI_DO_PRAGMA_(GCC warning #MSG)
#endif

#define __INIFORMAT_TABLE_CB_FIELDS__(NAME, OFFSET, SIZE, DEFVAL) \
    unsigned char NAME:SIZE;
#define __INIFORMAT_TABLE_CB_DEFAULT__(NAME, OFFSET, SIZE, DEFVAL) DEFVAL,
#define __INIFORMAT_TABLE_CB_ZERO__(NAME, OFFSET, SIZE, DEFVAL) 0,
#define _LIBCONFINI_INIFORMAT_TYPE_ \
    struct IniFormat { INIFORMAT_TABLE_AS(__INIFORMAT_TABLE_CB_FIELDS__) }
#define _LIBCONFINI_DEFAULT_FORMAT_ \
    { INIFORMAT_TABLE_AS(__INIFORMAT_TABLE_CB_DEFAULT__) }
#define _LIBCONFINI_UNIXLIKE_FORMAT_ \
    { INIFORMAT_TABLE_AS(__INIFORMAT_TABLE_CB_ZERO__) }



/*  PUBLIC MACROS  */


/**
    @brief  Call a user-given macro (that accepts four arguments) for each row
            of the table
**/
/*
    NOTE: The following table and the order of its rows **define** (and link
    together) both the #IniFormat and #IniFormatNum data types declared in this
    header
*/
#define INIFORMAT_TABLE_AS(_____)                 /*  IniFormat table  *\

        NAME                      BIT  SIZE DEFAULT
                                                                      */\
 _____( delimiter_symbol,         0,   7,   INI_EQUALS                ) \
 _____( case_sensitive,           7,   1,   false                     )/*
                                                                      */\
 _____( semicolon_marker,         8,   2,   INI_DISABLED_OR_COMMENT   ) \
 _____( hash_marker,              10,  2,   INI_DISABLED_OR_COMMENT   ) \
 _____( section_paths,            12,  2,   INI_ABSOLUTE_AND_RELATIVE ) \
 _____( multiline_nodes,          14,  2,   INI_MULTILINE_EVERYWHERE  )/*
                                                                      */\
 _____( no_single_quotes,         16,  1,   false                     ) \
 _____( no_double_quotes,         17,  1,   false                     ) \
 _____( no_spaces_in_names,       18,  1,   false                     ) \
 _____( implicit_is_not_empty,    19,  1,   false                     ) \
 _____( do_not_collapse_values,   20,  1,   false                     ) \
 _____( preserve_empty_quotes,    21,  1,   false                     ) \
 _____( disabled_after_space,     22,  1,   false                     ) \
 _____( disabled_can_be_implicit, 23,  1,   false                     )



/**
    @brief  Check whether a format does _not_ support escape sequences
**/
#define INIFORMAT_HAS_NO_ESC(FORMAT) \
    (FORMAT.multiline_nodes == INI_NO_MULTILINE && \
    FORMAT.no_double_quotes && FORMAT.no_single_quotes)


/**
    @brief  Check whether a given `char *` data type points to the global
            variable #INI_GLOBAL_IMPLICIT_VALUE or to any fragment of it
**/
#define INI_IS_IMPLICIT_SUBSTR(CHAR_PTR) \
    (CHAR_PTR >= INI_GLOBAL_IMPLICIT_VALUE && CHAR_PTR <= \
    INI_GLOBAL_IMPLICIT_VALUE + INI_GLOBAL_IMPLICIT_V_LEN)



/*  PUBLIC TYPEDEFS  */


/**
    @brief  24-bit bitfield representing the format of an INI file (INI
            dialect)
**/
typedef _LIBCONFINI_INIFORMAT_TYPE_ IniFormat;


/**
    @brief  Global statistics about an INI file
**/
typedef struct IniStatistics {
    const IniFormat format;
    const size_t bytes;
    const size_t members;
} IniStatistics;


/**
    @brief  Dispatch of a single INI node
**/
typedef struct IniDispatch {
    const IniFormat format;
    uint_least8_t type;
    char * data;
    char * value;
    const char * append_to;
    size_t d_len;
    size_t v_len;
    size_t at_len;
    size_t dispatch_id;
} IniDispatch;


/**
    @brief  The unique ID of an INI format (24-bit maximum)
**/
typedef uint_least32_t IniFormatNum;


/**
    @brief  Callback function for handling an #IniStatistics structure
**/
typedef int (* IniStatsHandler) (
    IniStatistics * statistics,
    void * user_data
);


/**
    @brief  Callback function for handling an #IniDispatch structure
**/
typedef int (* IniDispHandler) (
    IniDispatch * dispatch,
    void * user_data
);


/**
    @brief  Callback function for handling an INI string belonging to a
            sequence of INI strings
**/
typedef int (* IniStrHandler) (
    char * ini_string,
    size_t string_length,
    size_t string_num,
    IniFormat format,
    void * user_data
);


/**
    @brief  Callback function for handling a selected fragment of an INI string
**/
typedef int (* IniSubstrHandler) (
    const char * ini_string,
    size_t fragm_offset,
    size_t fragm_length,
    size_t fragm_num,
    IniFormat format,
    void * user_data
);



/*  PUBLIC FUNCTIONS  */


extern int strip_ini_cache (
    register char * const ini_source,
    const size_t ini_length,
    const IniFormat format,
    const IniStatsHandler f_init,
    const IniDispHandler f_foreach,
    void * const user_data
);


extern int load_ini_file (
    FILE * const ini_file,
    const IniFormat format,
    const IniStatsHandler f_init,
    const IniDispHandler f_foreach,
    void * const user_data
);


extern int load_ini_path (
    const char * const path,
    const IniFormat format,
    const IniStatsHandler f_init,
    const IniDispHandler f_foreach,
    void * const user_data
);


extern bool ini_string_match_ss (
    const char * const simple_string_a,
    const char * const simple_string_b,
    const IniFormat format
);


extern bool ini_string_match_si (
    const char * const simple_string,
    const char * const ini_string,
    const IniFormat format
);


extern bool ini_string_match_ii (
    const char * const ini_string_a,
    const char * const ini_string_b,
    const IniFormat format
);


extern bool ini_array_match (
    const char * const ini_string_a,
    const char * const ini_string_b,
    const char delimiter,
    const IniFormat format
);


extern size_t ini_unquote (
    char * const ini_string,
    const IniFormat format
);


extern size_t ini_string_parse (
    char * const ini_string,
    const IniFormat format
);


extern size_t ini_array_get_length (
    const char * const ini_string,
    const char delimiter,
    const IniFormat format
);


extern int ini_array_foreach (
    const char * const ini_string,
    const char delimiter,
    const IniFormat format,
    const IniSubstrHandler f_foreach,
    void * const user_data
);


extern size_t ini_array_shift (
    const char ** const ini_strptr,
    const char delimiter,
    const IniFormat format
);


extern size_t ini_array_collapse (
    char * const ini_string,
    const char delimiter,
    const IniFormat format
);


extern char * ini_array_break (
    char * const ini_string,
    const char delimiter,
    const IniFormat format
);


extern char * ini_array_release (
    char ** const ini_strptr,
    const char delimiter,
    const IniFormat format
);


extern int ini_array_split (
    char * const ini_string,
    const char delimiter,
    const IniFormat format,
    const IniStrHandler f_foreach,
    void * const user_data
);


extern void ini_global_set_lowercase_mode (
    const bool lowercase
);
#define ini_global_set_lowercase_mode \
    _LIBCONFINI_WARNING_("deprecated function `ini_global_set_lowercase_mode()`") \
    ini_global_set_lowercase_mode


extern void ini_global_set_implicit_value (
    char * const implicit_value,
    const size_t implicit_v_len
);


extern IniFormatNum ini_fton (
    const IniFormat format
);


extern IniFormat ini_ntof (
    const IniFormatNum format_id
);


extern int ini_get_bool (
    const char * const simple_string,
    const int when_fail
);


extern int ini_get_bool_i (
    const char * const ini_string,
    const int when_fail,
    const IniFormat format
);



/*  PUBLIC LINKS  */


extern int (* const ini_get_int) (
    const char * ini_string
);


extern long int (* const ini_get_lint) (
    const char * ini_string
);


extern long long int (* const ini_get_llint) (
    const char * ini_string
);


extern double (* const ini_get_double) (
    const char * ini_string
);


/*
    Legacy support -- please **do not use these functions**!
*/
#define ini_get_float \
    _LIBCONFINI_WARNING_("function `ini_get_float()` is deprecated for parsing a `double` data type; use `ini_get_double()` instead") \
    ini_get_double



/*  PUBLIC CONSTANTS AND VARIABLES  */


/**
    @brief  Error mask (flags not present in user-generated interruptions)
**/
#define CONFINI_ERROR 252


/**
    @brief  Error codes
**/
enum ConfiniInterruptNo {
    CONFINI_SUCCESS = 0,    /**< There have been no interruptions, everything
                                 went well [value=0] **/
    CONFINI_IINTR = 1,      /**< Interrupted by the user during `f_init()`
                                 [value=1] **/
    CONFINI_FEINTR = 2,     /**< Interrupted by the user during `f_foreach()`
                                 [value=2] **/
    CONFINI_ENOENT = 4,     /**< File inaccessible [value=4] **/
    CONFINI_ENOMEM = 5,     /**< Error allocating virtual memory [value=5] **/
    CONFINI_EIO = 6,        /**< Error reading the file [value=6] **/
    CONFINI_EOOR = 7,       /**< Out-of-range error: callbacks are more than
                                 expected [value=7] **/
    CONFINI_EBADF = 8,      /**< The stream specified is not a seekable stream
                                 [value=8] **/
    CONFINI_EFBIG = 9,      /**< File too large [value=9] **/
    CONFINI_EROADDR = 10    /**< Address is read-only [value=10] **/
};


/**
    @brief  Disabled flag (i.e. third bit, present only in non-active node
            types)
**/
#define INI_DISABLED_FLAG 4


/**
    @brief  INI node types and possible values of #IniDispatch::type

    See also #INI_DISABLED_FLAG.
**/
enum IniNodeType {
    INI_UNKNOWN = 0,            /**< This is a node impossible to categorize
                                     [value=0] **/
    INI_VALUE = 1,              /**< Not used by **libconfini** (values are
                                     dispatched together with keys) -- but
                                     available for user's implementations
                                     [value=1] **/
    INI_KEY = 2,                /**< This is a key [value=2] **/
    INI_SECTION = 3,            /**< This is a section path [value=3] **/
    INI_COMMENT = 4,            /**< This is a comment [value=4] **/
    INI_INLINE_COMMENT = 5,     /**< This is an inline comment [value=5] **/
    INI_DISABLED_KEY = 6,       /**< This is a disabled key [value=6] **/
    INI_DISABLED_SECTION = 7    /**< This is a disabled section path
                                     [value=7] **/
};


/**
    @brief  Common array and key-value delimiters (but a delimiter may also be
            any other ASCII character not present in this list)
**/
enum IniDelimiters {
    INI_ANY_SPACE = 0,  /**< In multi-line INIs:
                             `/(?:\\(?:\n\r?|\r\n?)|[\t \v\f])+/`, in
                             non-multi-line INIs: `/[\t \v\f])+/` **/
    INI_EQUALS = '=',   /**< Equals character (`=`) **/
    INI_COLON = ':',    /**< Colon character (`:`) **/
    INI_DOT = '.',      /**< Dot character (`.`) **/
    INI_COMMA = ','     /**< Comma character (`,`) **/
};


/**
    @brief  Possible values of #IniFormat::semicolon_marker and
            #IniFormat::hash_marker (i.e., meaning of `/\s+;/` and `/\s+#/` in
            respect to a format)
**/
enum IniCommentMarker {
    INI_DISABLED_OR_COMMENT = 0,    /**< This marker opens a comment or a
                                         disabled entry **/
    INI_ONLY_COMMENT = 1,           /**< This marker opens a comment **/
    INI_IGNORE = 2,                 /**< This marker opens a comment that has
                                         been marked for deletion and must not
                                         be dispatched or counted **/
    INI_IS_NOT_A_MARKER = 3         /**< This is not a marker at all, but a
                                         normal character instead **/
};


/**
    @brief  Possible values of #IniFormat::section_paths
**/
enum IniSectionPaths {
    INI_ABSOLUTE_AND_RELATIVE = 0,  /**< Section paths starting with a dot
                                         express nesting to the current parent,
                                         to root otherwise **/
    INI_ABSOLUTE_ONLY = 1,          /**< Section paths starting with a dot will
                                         be cleaned of their leading dot and
                                         appended to root **/
    INI_ONE_LEVEL_ONLY = 2,         /**< Format supports sections, but the dot
                                         does not express nesting and is not a
                                         meta-character **/
    INI_NO_SECTIONS = 3             /**< Format does *not* support sections --
                                         `/\[[^\]]*\]/g`, if any, will be
                                         treated as keys! **/
};


/**
    @brief  Possible values of #IniFormat::multiline_nodes
**/
enum IniMultiline {
    INI_MULTILINE_EVERYWHERE = 0,       /**< Comments, section paths and keys
                                             -- disabled or not -- are allowed
                                             to be multi-line **/
    INI_BUT_COMMENTS = 1,               /**< Only section paths and keys --
                                             disabled or not -- are allowed to
                                             be multi-line **/
    INI_BUT_DISABLED_AND_COMMENTS = 2,  /**< Only active section paths and
                                             active keys are allowed to be
                                             multi-line **/
    INI_NO_MULTILINE = 3                /**< Multi-line escape sequences are
                                             disabled **/
};


/**
    @brief  A model format for standard INI files
**/
static const IniFormat INI_DEFAULT_FORMAT = _LIBCONFINI_DEFAULT_FORMAT_;


/**
    @brief  A model format for Unix-like .conf files (where space characters
            are delimiters between keys and values)
**/
/*  All fields are set to `0` here.  */
static const IniFormat INI_UNIXLIKE_FORMAT = _LIBCONFINI_UNIXLIKE_FORMAT_;


/**

    @brief          If set to `true`, key and section names in case-insensitive
                    INI formats will be dispatched lowercase, verbatim
                    otherwise (default value: `false`)
    @deprecated     Deprecated since version 1.15.0 (it will be removed in
                    version 2.0.0)
**/
extern bool INI_GLOBAL_LOWERCASE_MODE;
#define INI_GLOBAL_LOWERCASE_MODE \
    _LIBCONFINI_WARNING_("global variable `INI_GLOBAL_LOWERCASE_MODE` is deprecated") \
    INI_GLOBAL_LOWERCASE_MODE


/**
    @brief  Value to be assigned to implicit keys (default value: `NULL`)
**/
extern char * INI_GLOBAL_IMPLICIT_VALUE;


/**
    @brief  Length of the value assigned to implicit keys (default value: `0`)
**/
extern size_t INI_GLOBAL_IMPLICIT_V_LEN;



/*  CLEAN THE PRIVATE ENVIRONMENT  */


#undef _LIBCONFINI_UNIXLIKE_FORMAT_
#undef _LIBCONFINI_DEFAULT_FORMAT_
#undef _LIBCONFINI_INIFORMAT_TYPE_
#undef __INIFORMAT_TABLE_CB_ZERO__
#undef __INIFORMAT_TABLE_CB_DEFAULT__
#undef __INIFORMAT_TABLE_CB_FIELDS__



/*  END OF `__CONFINI_H__`  */


#ifdef __cplusplus
}
#endif


#endif


/*  EOF  */

